src/Bicep.Core/Registry/AzureContainerRegistryManager.cs

```diff
@@ -27,104 +35,297 @@ public class AzureContainerRegistryManager
+                        "digest": "sha256:...",
+                        "size": 2774
+                      }
+                    ],
+                    "annotations": {
+                      "org.opencontainers.image.description": "module description"
+                      "org.opencontainers.image.documentation": "https://www.contoso.com/moduledocumentation.html"
+                    }
+                  }
+             */
 
+            var annotations = new Dictionary<string, string>();
             if (!string.IsNullOrWhiteSpace(documentationUri))
             {
                 annotations[LanguageConstants.OciOpenContainerImageDocumentationAnnotation] = documentationUri;
             }
-
             if (!string.IsNullOrWhiteSpace(description))
             {
                 annotations[LanguageConstants.OciOpenContainerImageDescriptionAnnotation] = description;
             }
 
-            var manifest = new OciManifest(2, artifactType, configDescriptor, layerDescriptors, annotations);
+            // Adding a timestamp is important to ensure any sources manifests will always point to a unique module
+            //   manifest, even if something in the sources changes that doesn't affect the compiled output (in which
+            //   case, it we don't delete the previous source referrer, you will have two source manifests pointing
+            //   to the same module manifest).
+            // And it can be useful for its own sake as well.
+            // TODO: This currently confuses the tests.  Do we need to delete the previous source manifests and previous
+            //   module manifest?
+            //annotations[LanguageConstants.OciOpenContainerImageCreatedAnnotation] = timestamp;
+
+            var moduleManifestDescriptor = await PushArtifactAsync(
+                blobClient,
+                moduleReference.Tag,
+                subject: null,
+                artifactType,
+                config,
+                annotations,
+                layers);
+            Trace.WriteLine($"Uploaded module {moduleReference.FullyQualifiedReference} with digest {moduleManifestDescriptor.Digest}");
+
+            return moduleManifestDescriptor;
+        }
+
+        private async Task<OciDescriptor> PushSourceManifestAsync(IOciRegistryContentClient blobClient, OciArtifactModuleReference moduleReference, OciDescriptor moduleManifest, Stream bicepSources, string timestamp)
+        {
+            // See more information here: https://learn.microsoft.com/en-us/azure/container-registry/container-registry-oci-artifacts
+
+            /* Sample source manifest:
+                {
+                   "schemaVersion": 2,
+                   "artifactType":"application/vnd.ms.bicep.module.source",
+                   "config": {
+                       "mediaType":"application/vnd.ms.bicep.module.source.config.v1+json",
+                       "digest":"sha256:...",
+                       "size": 2
```

Comment:
> "size": 2 [](http://example.com/codeflow?start=22&length=10)

Are we writing config blobs?

Reply:
Yes, see this comment:

             // ACR won't recognize this as a valid attachment unless this is valid JSON, so write out "{}" as the configuration content
            var config = new StreamDescriptor(
                new MemoryStream(Encoding.UTF8.GetBytes("{}")),
                BicepMediaTypes.BicepSourceConfigV1);

I think this may be a bug on ACR's part, but not positive.

I actually think we should be writing "{}" for our module manifests, see here: https://github.com/opencontainers/image-spec/blob/main/manifest.md#guidance-for-an-empty-descriptor